providers/asrockrack/firmware.go

```diff
@@ -0,0 +1,231 @@
+	switch component {
+	case devices.SlugBIOS:
+		return a.firmwareUpdateBIOSStatus(ctx, installVersion)
+	case devices.SlugBMC:
+		return a.firmwareUpdateBMCStatus(ctx, installVersion)
+	default:
+		return "", errors.Wrap(bmclibErrs.ErrFirmwareInstall, "component unsupported: "+component)
+	}
+}
+
+// firmwareInstallBMC uploads and installs firmware for the BMC component
+func (a *ASRockRack) firmwareInstallBMC(ctx context.Context, reader io.Reader, fileSize int64) error {
+	var err error
+
+	// 1. set the device to flash mode - prepares the flash
+	a.log.V(2).Info("info", "action", "set device to flash mode, takes a minute...", "step", "1/4")
+	err = a.setFlashMode(ctx)
+	if err != nil {
+		return fmt.Errorf("failed in step 1/4 - set device to flash mode: " + err.Error())
+	}
+
+	// 2. upload firmware image file
+	a.log.V(2).Info("info", "action", "upload BMC firmware image", "step", "2/4")
+	err = a.uploadFirmware(ctx, "api/maintenance/firmware", reader, fileSize)
+	if err != nil {
+		return fmt.Errorf("failed in step 2/4 - upload BMC firmware image: " + err.Error())
+	}
+
+	// 3. BMC to verify the uploaded file
+	err = a.verifyUploadedFirmware(ctx)
+	a.log.V(2).Info("info", "action", "BMC verify uploaded firmware", "step", "3/4")
+	if err != nil {
+		return fmt.Errorf("failed in step 3/4 - BMC verify uploaded firmware: " + err.Error())
+	}
+
+	// 4. Run the upgrade - preserving current config
+	a.log.V(2).Info("info", "action", "proceed with upgrade, preserve current configuration", "step", "4/4")
+	err = a.upgradeBMC(ctx)
+	if err != nil {
+		return fmt.Errorf("failed in step 4/4 - proceed with upgrade: " + err.Error())
+	}
+
+	return nil
+}
+
+// firmwareInstallBIOS uploads and installs firmware for the BIOS component
+func (a *ASRockRack) firmwareInstallBIOS(ctx context.Context, reader io.Reader, fileSize int64) error {
+	var err error
+
+	// 1. upload firmware image file
+	a.log.V(2).Info("info", "action", "upload BIOS firmware image", "step", "1/3")
+	err = a.uploadFirmware(ctx, "api/asrr/maintenance/BIOS/firmware", reader, fileSize)
+	if err != nil {
+		return fmt.Errorf("failed in step 1/3 - upload firmware image: " + err.Error())
+	}
+
+	// 2. set update parameters to preserve configurations
+	a.log.V(2).Info("info", "action", "set flash configuration", "step", "2/3")
+	err = a.biosUpgradeConfiguration(ctx)
+	if err != nil {
+		return fmt.Errorf("failed in step 2/3 - set flash configuration: " + err.Error())
+	}
+
+	// 3. run upgrade
+	a.log.V(2).Info("info", "action", "proceed with upgrade", "step", "3/3")
+	err = a.upgradeBIOS(ctx)
+	if err != nil {
+		return fmt.Errorf("failed in step 3/3 - proceed with upgrade: " + err.Error())
+	}
+
+	return nil
+}
+
+// firmwareUpdateBMCStatus returns the BMC firmware install status
+func (a *ASRockRack) firmwareUpdateBMCStatus(ctx context.Context, installVersion string) (status string, err error) {
+	os.Setenv("BMCLIB_LOG_LEVEL", "trace")
```

Comment:
Why are trace logs necessary?

Reply:
I did not clean up properly there, fixed in latest commit